src/inmanta/agent/executor.py

```diff
@@ -369,7 +608,46 @@ async def get_environment(self, blueprint: EnvBlueprint, threadpool: ThreadPoolE
                     blueprint.blueprint_hash(),
                 )
                 return self._environment_map[blueprint]
-            return await self.create_environment(blueprint, threadpool)
+            # The whole creation of the virtual environment is under lock. Therefore, we know that it will not race with the
+            # cleanup as the `INMANTA_ENV_STATUS_FILENAME` will be touched at the end of the creation.
+            return await self.create_environment(blueprint)
+
+    async def get_pool_members(self) -> Sequence[ExecutorVirtualEnvironment]:
+        """
+        Retrieve the members of the pool: those that need to be checked for the cleanup.
+        """
+        # We should walk once for the first-level of folders!
+        envs_dir = pathlib.Path(self.envs_dir)
+        try:
+            _, root_folders, _ = next(os.walk(self.envs_dir))
+        except StopIteration as e:
+            raise RuntimeError(f"An error occurred while checking following directory: `{envs_dir.absolute()}`!") from e
+
+        venv_path_to_blueprint = {pathlib.Path(v.env_path): k for k, v in self._environment_map.items()}
+
+        pool_members = []
+        for folder in root_folders:
+            # the name of the folder is the hash!
+            async with self._locks.get(folder):
+                current_folder = envs_dir / folder
+                blueprint = venv_path_to_blueprint.get(current_folder, None)
+                if blueprint is not None and blueprint in self._environment_map:
+                    current_executor_environment = self._environment_map[blueprint]
+                else:
+                    current_executor_environment = ExecutorVirtualEnvironment(env_path=str(current_folder.absolute()))
+
+                pool_members.append(current_executor_environment)
```

Comment:
I wonder if we shouldn't cache all this: load from disk when initializing, then work from memory?